packages/grid/x-data-grid-pro/src/utils/tree/removeDataRowFromTree.ts

```diff
@@ -0,0 +1,124 @@
+  GridTreeNode,
+} from '@mui/x-data-grid';
+import { GridTreeDepths, GridRowTreeUpdatedGroups } from '@mui/x-data-grid/internals';
+import {
+  getNodePathInTree,
+  getGroupRowIdFromPath,
+  removeNodeFromTree,
+  updateGroupNodeIdAndAutoGenerated,
+  addActionToGroup,
+} from './utils';
+
+interface RemoveDataRowFromTreeParams {
+  /**
+   * ID of the data row to remove from the tree.
+   */
+  id: GridRowId;
+  /**
+   * Tree from which to remove the data row.
+   * This tree can be mutated but it's children should not.
+   * For instance:
+   *
+   * - `tree[nodeId] = newNode` => valid
+   * - `tree[nodeId].children.push(newNodeId)` => invalid
+   */
+  tree: GridRowTreeConfig;
+  /**
+   * Amount of nodes at each depth of the tree.
+   * This object can be mutated.
+   * For instance:
+   *
+   * - `treeDepths[nodeDepth] = treeDepth[nodeDepth] + 1` => valid
+   */
+  treeDepths: GridTreeDepths;
+  /**
+   * Object tracking the action performed on each group.
+   * Used to decide which groups to refresh on sorting, filtering, aggregation, ...
+   * If not defined, then the tracking will be skipped.
+   */
+  updatedGroups?: GridRowTreeUpdatedGroups;
+}
+
+const removeNodeAndCleanParent = ({
+  node,
+  tree,
+  treeDepths,
+  updatedGroups,
+}: {
+  node: GridTreeNode;
+  tree: GridRowTreeConfig;
+  treeDepths: GridTreeDepths;
+  updatedGroups?: GridRowTreeUpdatedGroups;
+}) => {
+  removeNodeFromTree({
+    node,
+    tree,
+    treeDepths,
+  });
+
+  if (node.type === 'group' && node.footerId != null) {
+    removeNodeFromTree({
+      node: tree[node.footerId],
+      tree,
+      treeDepths,
+    });
+  }
+
+  const parentNode = tree[node.parent!] as GridGroupNode;
+
+  addActionToGroup(updatedGroups, parentNode.id, 'removeChildren');
+
+  const shouldDeleteGroup =
+    parentNode.id !== GRID_ROOT_GROUP_ID && parentNode.children.length === 0;
+  if (shouldDeleteGroup) {
+    if (parentNode.isAutoGenerated) {
+      removeNodeAndCleanParent({ node, tree, treeDepths });
+    } else {
+      tree[parentNode.id] = {
+        type: 'leaf',
+        id: parentNode.id,
+        depth: parentNode.depth,
+        parent: parentNode.parent!,
+        groupingKey: parentNode.groupingKey,
+      };
+    }
+  }
+};
+
+/**
+ * Removed a data row from the tree.
+ * If the node is a non-empty group, replace it with an auto-generated group to be able to keep displaying its children.
+ * If not, remove it and recursively clean its parent with the following rules:
+ * - An empty auto-generated should be removed from the tree
+ * - An empty non-auto-generated should be turned into a leaf
+ */
+export const removeDataRowFromTree = ({
+  id,
+  tree,
+  treeDepths,
+  updatedGroups,
+}: RemoveDataRowFromTreeParams) => {
+  const node = tree[id];
+
+  if (node.type === 'group' && node.children.length > 0) {
+    addActionToGroup(updatedGroups, node.parent!, 'removeChildren');
+    addActionToGroup(updatedGroups, node.parent!, 'insertChildren');
```

Comment:
Why isn't it only `insertChildren`in this conditional block, and `removeChildren` in the `else` block?

Reply:
The  `else` block has it own `removeChildren` calls. It calls a function that can remove / add or mutate groups.

And the `if` deletes a group and create a new one (I consider that changing the ID of a group is like deleting a group and creating one, that could be considered once we really use those actions)

It would probably be more explicit if the content of the `if` want a standalone function like the content of the `else`.